### ISSUE TITLE
lookup plugin code is run multiple times for each template task when lookups are used in the inventory

### ISSUE THREAD

**Issue by Ernest0x** [ansible/ansible#9623]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.8 (devel d36c38c35e) last updated 2014/11/25 14:09:58 (GMT +300)
  lib/ansible/modules/core: (detached HEAD 10ebcccedb) last updated 2014/11/25 14:10:20 (GMT +300)
  lib/ansible/modules/extras: (detached HEAD 317654dba5) last updated 2014/11/25 14:10:21 (GMT +300)
  v2/ansible/modules/core: (detached HEAD cb69744bce) last updated 2014/11/24 21:06:56 (GMT +300)
  v2/ansible/modules/extras: (detached HEAD 8a4f07eecd) last updated 2014/11/24 21:07:01 (GMT +300)
  configured module search path = /usr/share/ansible
##### Environment:

Archlinux
##### Summary:

If there are variables in the inventory with lookups, e.g:

```
test_var: "{{ lookup('testlookup', '') }}"
```

The code of 'testlookup' plugin is triggered multiple times for each 'template' task, even if the template does not include {{ test_var }} 
##### Steps To Reproduce:

Write a test lookup that prints something to indicate whenever it is running:

```
 import ansible.utils as utils

 class LookupModule(object):

     def __init__(self, basedir=None, **kwargs):
         self.basedir = basedir

     def run(self, terms, inject=None, **kwargs):
         print('I am running')
         return []
```

Write a playbook that runs a template task to render {{ test_var }}, as well as many other template tasks that do not render {{ test_var }}. See how many time 'I am running' is printed
##### Expected Results:

'I am running' should be printed only once
##### Actual Results:

'I am running' is printed multiple times.
##### Further notes:

I was actually experiencing another issue, which lead me to the abnormal behavior above. The issue that lead me to the above, has to do with a lookup plugin I have developed (and works on 1.7.2) that passes the lookup's terms to utils.template.template_from_string() to render a variable that exists in a hostvar file (encrypted with ansible-vault). While that variable does exist in the inventory, template_from_string() fails with 'undefined'. So, I suspect that something really bad is happening when trying, in a template, to access a variable from the inventory, which calls a lookup plugin, whose code in turn tries to access another (vaulted) variable from the inventory. I think that it is possible that there is a common root for both of these issues. Was there any core change in ansible's templating mechanism for 1.8?

**Comment by mpdehaan** [comment]
Lookup plugins are lazy evaluated and this may occur at multiple points in time.  If you are doing something dynamic in inventory, it is probable that a dynamic inventory script may be a better option.

In the v2 tree these will be evaluated much less, so I'm going to close this since the new architecture will track the resolution of this item.

Thanks!

**Comment by drzraf** [comment]
Any progress on this one?
It's a huge drawback. If the lookup(pipe) is a slow process, the uncached lazy evaluation will practically makes the playbook unusable.
In the variables of my playbook I initialize two variables using a pipe, each of them taking 3 seconds to run. But... since my playbook use 3 roles, a couple of template, ... ... this slow process is run more than 50 times which is just impracticable.

There must be a syntax to state that given initialization just have to takes place **now** and once for all (what would be logical after all)

**Comment by rouge8** [comment]
If anyone else runs into this, I've created a cached lookup plugin that lets you cache any lookup, e.g.:

```yaml
test_var: "{{ lookup('cached', 'testlookup', '') }}"
first_date: "{{ lookup('cached', 'pipe', 'date') }}"
```

https://github.com/GoodRx/ansible-cached-lookup
